go/pkg/command/command.go

```diff
@@ -0,0 +1,273 @@
+	for _, i := range s {
+		*buf = append(*buf, []byte(i)...)
+	}
+}
+
+func marshallSortedSlice(s []string, buf *[]byte) {
+	ss := make([]string, len(s))
+	copy(ss, s)
+	sort.Strings(ss)
+	marshallSlice(ss, buf)
+}
+
+// Validate checks whether all required command fields have been specified.
+func (c *Command) Validate() error {
+	if c.Args == nil {
+		return errors.New("missing command arguments")
+	}
+	if c.ExecRoot == "" {
+		return errors.New("missing command exec root")
+	}
+	if c.InputSpec == nil {
+		return errors.New("missing command input spec")
+	}
+	if c.Identifiers == nil {
+		return errors.New("missing command identifiers")
+	}
+	return nil
+}
+
+// Generates a stable id for the command.
+func (c *Command) stableId() string {
+	var buf []byte
+	marshallSlice(c.Args, &buf)
+	buf = append(buf, []byte(c.ExecRoot)...)
+	buf = append(buf, []byte(c.WorkingDir)...)
+	marshallSortedSlice(c.OutputFiles, &buf)
+	marshallSortedSlice(c.OutputDirs, &buf)
+	buf = append(buf, []byte(c.Timeout.String())...)
+	marshallMap(c.Platform, &buf)
+	if c.InputSpec != nil {
+		marshallMap(c.InputSpec.EnvironmentVariables, &buf)
+		marshallSortedSlice(c.InputSpec.Inputs, &buf)
+		inputExclusions := make([]*InputExclusion, len(c.InputSpec.InputExclusions))
+		copy(inputExclusions, c.InputSpec.InputExclusions)
+		sort.Slice(inputExclusions, func(i, j int) bool {
+			e1 := inputExclusions[i]
+			e2 := inputExclusions[j]
+			return e1.Regex > e2.Regex || e1.Regex == e2.Regex && e1.Kind > e2.Kind
+		})
+		for _, e := range inputExclusions {
+			buf = append(buf, []byte(e.Regex)...)
+			buf = append(buf, []byte(e.Kind.String())...)
+		}
+	}
+	sha256Arr := sha256.Sum256(buf)
+	return hex.EncodeToString(sha256Arr[:])[:8]
+}
+
+// FillDefaultFieldValues initializes valid default values to inner Command fields.
+// This function should be called on every new Command object before use.
+func (c *Command) FillDefaultFieldValues() {
+	if c == nil {
+		return
+	}
+	if c.Identifiers == nil {
+		c.Identifiers = &Identifiers{}
+	}
+	if c.Identifiers.CommandId == "" {
+		c.Identifiers.CommandId = c.stableId()
+	}
+	if c.Identifiers.ToolName == "" {
+		c.Identifiers.ToolName = "remote-client"
+	}
+	if c.Identifiers.InvocationId == "" {
+		c.Identifiers.InvocationId = uuid.New()
+	}
+	if c.InputSpec == nil {
+		c.InputSpec = &InputSpec{}
+	}
+}
+
+// ExecutionOptions specify how to execute a given Command.
+type ExecutionOptions struct {
+	// Whether to accept remotely cached action results. Defaults to true.
+	AcceptCached bool
+
+	// When set, this execution results will not be cached remotely.
+	DoNotCache bool
+
+	// Download command outputs after execution. Defaults to true.
+	DownloadOutputs bool
+}
+
+// DefaultExecutionOptions returns the recommended ExecutionOptions.
+func DefaultExecutionOptions() *ExecutionOptions {
+	return &ExecutionOptions{
+		AcceptCached:    true,
+		DoNotCache:      false,
+		DownloadOutputs: true,
+	}
+}
+
```

Comment:
Shouldn't this be part of the sdk not here?

Reply:
Not sure, because these options are also relevant to local execution. Same question about Metadata and Result. I suggest to start by exporting these here, and see if we later want to split them out.

Reply:
sgtm